docs/en/modules/sjson.md

```diff
@@ -0,0 +1,218 @@
+
+####Returns
+A string of up to `size` bytes, or `nil` if the encoding is complete and all data has been returned.
+
+#### Example
+The following example prints out (in 64 byte chunks) a JSON encoded string containing the first 4k of every file in the file system. The total string
+can be bigger than the total amount of memory on the NodeMCU.
+```
+function files() 
+   result = {}
+   for k,v in pairs(file.list()) do
+     result[k] = function() return file.open(k):read(4096) end
+   end
+   return result
+end
+
+local encoder = sjson.encoder(files())
+
+while true do
+   data = encoder:read(64)
+   if not data then
+      break
+   end
+   print(data)
+end
+```
+
+## sjson.encode()
+
+Encode a Lua table to a JSON string. This is a convenience method provided for backwards compatibility with `cjson`.
+
+####Syntax
+`sjson.encode(table [, opts])`
+
+####Parameters
+- `table` data to encode
+- `opts` an optional table of options. The possible entries are:
+    - `depth` the maximum encoding depth needed to encode the table. The default is 20 which should be enough for nearly all situations.
+    - `null` the string value to treat as null.
+
+####Returns
+JSON string
+
+####Example
+```lua
+ok, json = pcall(sjson.encode, {key="value"})
+if ok then
+  print(json)
+else
+  print("failed to encode!")
+end
+```
+
+## sjson.decoder()
+
+This makes a decoder object that can parse a JSON encoded string into a lua object. A metatable can be specified for all the newly created lua tables. This allows
+you to handle each value as it is inserted into each table (by implementing the `__newindex` method).
+
+####Syntax
+`sjson.decoder([opts])`
+
+#### Parameters
+- `opts` an optional table of options. The possible entries are:
+    - `depth` the maximum encoding depth needed to encode the table. The default is 20 which should be enough for nearly all situations.
+    - `null` the string value to treat as null.
+    - `metatable` a table to use as the metatable for all the new tables in the returned object.
+
+#### Returns
+A `sjson.decoder` object
+
+####Metatable
+
+There are two principal methods that are invoked in the metatable (if it is present).
+
+- `__newindex` this is the standard method invoked whenever a new table element is created.
+- `setpath` this is invoked (if defined) whenever a new table is created. It is invoked with two arguments:
+    - `table` this is the newly created table
+    - `path` this is a list of the keys from the root. 
+
+For example, when decoding `{ "foo": [1, 2, []] }` the setpath will be invoked as follows:
+
+- `setpath({}, {})` the `table` argument is the object that will correspond with the value of the JSON object.
+- `setpath({}, {"foo"})` the `table` argument is the object that will correspond with the value of the outer JSON array.
+- `setpath({}, {"foo", 3})` the `table` argument is the object that will correspond to the empty inner JSON array.
```

Comment:
I haven't really thought through what the likely usecases are here, but I suspect a `checkpath` would be more readily understood. Or it could be that we just need some more docs or examples?